File: .github/builder/generate-build-matrix.js
const core = require("@actions/core");
const axios = require("axios");
const cheerio = require("cheerio");

// Extract directory names from an Apache-style index page.
// Keeps only relative single-segment hrefs (e.g. "ath79/"), drops "../" and absolute paths.
function parseLinks($) {
  return $('a[href$="/"]')
    .map((i, el) => $(el).attr("href"))
    .get()
    .filter((h) => h && !h.startsWith("/") && !h.includes("://") && h !== "../")
    .map((h) => h.replace(/\/$/, ""));
}

// Fetch a URL and return the response data, or null on non-2xx / network error.
async function fetchSafe(url) {
  try {
    const res = await axios.get(url, { validateStatus: () => true });
    if (res.status < 200 || res.status >= 300) {
      console.warn(`SKIP ${url} (HTTP ${res.status})`);
      return null;
    }
    return res.data;
  } catch (err) {
    console.warn(`SKIP ${url} (${err.message})`);
    return null;
  }
}

async function generateMatrix(version, filterTargets, filterSubtargets) {
  const versionURL = `https://downloads.openwrt.org/releases/${version}/targets/`;

  const indexData = await fetchSafe(versionURL);
  if (!indexData) {
    core.setFailed(`OpenWrt release not found: ${versionURL}`);
    return;
  }

  const $ = cheerio.load(indexData);
  const allTargets = parseLinks($);

  const builds = [];

  for (const target of allTargets) {
    // Apply target filter early to avoid unnecessary requests
    if (filterTargets.length && !filterTargets.includes(target)) continue;

    const targetData = await fetchSafe(`${versionURL}${target}/`);
    if (!targetData) continue;

    const $t = cheerio.load(targetData);
    const subtargets = parseLinks($t);

    for (const subtarget of subtargets) {
      if (filterSubtargets.length && !filterSubtargets.includes(subtarget)) continue;

      const subtargetURL = `${versionURL}${target}/${subtarget}/`;
      const subtargetData = await fetchSafe(subtargetURL);
      if (!subtargetData) continue;

      const $st = cheerio.load(subtargetData);
      const hrefs = $st("a").map((i, el) => $st(el).attr("href")).get();

      // Skip if no SDK available for this target
      const sdkFile = hrefs.find((h) => h?.match(/openwrt-sdk-.*\.tar\.zst$/));
      if (!sdkFile) continue;

      // Get pkgarch from Packages.manifest
      let pkgarch = "";
      const manifestFile = hrefs.find((h) => h?.match(/Packages\.manifest/));
      if (manifestFile) {
        const manifest = await fetchSafe(`${subtargetURL}${manifestFile}`);
        if (manifest) {
          const m = manifest.match(/Architecture: (\S+)/);
          pkgarch = m ? m[1] : "";
        }
      }

      builds.push({ tag: version, target, subtarget, pkgarch, sdk_file: sdkFile });
    }
  }

  core.setOutput("job-config", JSON.stringify(builds));
  console.log(`Generated ${builds.length} build targets`);
}

const version = process.argv[2] || "24.10.4";
const targets = process.argv[3]?.split(",").filter(Boolean) || [];
const subtargets = process.argv[4]?.split(",").filter(Boolean) || [];
generateMatrix(version, targets, subtargets);

File: .github/builder/generate_build_matrix.py
#!/usr/bin/env python3

import json
import os
import re
import sys
import urllib.error
import urllib.request
from html.parser import HTMLParser


class LinkParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.hrefs = []

    def handle_starttag(self, tag, attrs):
        if tag != "a":
            return
        for key, value in attrs:
            if key == "href" and value:
                self.hrefs.append(value)
                break


def parse_links(html):
    parser = LinkParser()
    parser.feed(html)
    return [
        href.rstrip("/")
        for href in parser.hrefs
        if href.endswith("/")
        and not href.startswith("/")
        and "://" not in href
        and href != "../"
    ]


def fetch_safe(url):
    try:
        with urllib.request.urlopen(url) as response:
            return response.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as err:
        print(f"SKIP {url} (HTTP {err.code})", file=sys.stderr)
        return None
    except urllib.error.URLError as err:
        print(f"SKIP {url} ({err.reason})", file=sys.stderr)
        return None


def set_output(name, value):
    github_output = os.environ.get("GITHUB_OUTPUT")
    if github_output:
        with open(github_output, "a", encoding="utf-8") as fh:
            fh.write(f"{name}={value}\n")
    else:
        print(f"{name}={value}")


def main():
    version = sys.argv[1] if len(sys.argv) > 1 else "24.10.4"
    filter_targets = [item for item in (sys.argv[2] if len(sys.argv) > 2 else "").split(",") if item]
    filter_subtargets = [item for item in (sys.argv[3] if len(sys.argv) > 3 else "").split(",") if item]

    version_url = f"https://downloads.openwrt.org/releases/{version}/targets/"
    index_data = fetch_safe(version_url)
    if not index_data:
        print(f"OpenWrt release not found: {version_url}", file=sys.stderr)
        sys.exit(1)

    builds = []

    for target in parse_links(index_data):
        if filter_targets and target not in filter_targets:
            continue

        target_data = fetch_safe(f"{version_url}{target}/")
        if not target_data:
            continue

        for subtarget in parse_links(target_data):
            if filter_subtargets and subtarget not in filter_subtargets:
                continue

            subtarget_url = f"{version_url}{target}/{subtarget}/"
            subtarget_data = fetch_safe(subtarget_url)
            if not subtarget_data:
                continue

            hrefs = LinkParser()
            hrefs.feed(subtarget_data)
            href_list = hrefs.hrefs

            sdk_file = next((href for href in href_list if re.match(r"openwrt-sdk-.*\.tar\.zst$", href)), None)
            if not sdk_file:
                continue

            pkgarch = ""
            manifest_file = next((href for href in href_list if re.match(r"Packages\.manifest$", href)), None)
            if manifest_file:
                manifest = fetch_safe(f"{subtarget_url}{manifest_file}")
                if manifest:
                    match = re.search(r"Architecture: (\S+)", manifest)
                    if match:
                        pkgarch = match.group(1)

            builds.append(
                {
                    "tag": version,
                    "target": target,
                    "subtarget": subtarget,
                    "pkgarch": pkgarch,
                    "sdk_file": sdk_file,
                }
            )

    set_output("job-config", json.dumps(builds))
    print(f"Generated {len(builds)} build targets")


if __name__ == "__main__":
    main()
